### PR TITLE
Add an icon and visually hidden a11y text to links that open in a new tab

### LINKED ISSUE #1105: Links that open in a new tab/window need to inform users a new tab/window will open
There are some cases where links use `target="_blank"`. For example:

screen shot 2017-06-10 at 14 10 00

current link html:
`<a href="https://codex.wordpress.org/Excerpt" target="_blank">Learn more about manual excerpts</a>`

In all these cases, the link should inform users a new tab/window will open;
- for accessibility: some visually hidden text in the link text would be ok; core recently standardized to always use `(opens in a new window)`, see https://core.trac.wordpress.org/changeset/40643
- visually: I'll suggest to consider to use some visual indicator too, it's also a usability thing but this is definitely up to designers :)

See also: https://core.trac.wordpress.org/query?keywords=~target-blank

### PR DESCRIPTION
**Pull request by talldan**
## Description
Closes #1105 

Adds an icon and the visually hidden text '(Opens in a new window)' for improved a11y to two links that open in a new tab.

- Use `ExternalLink` component for the link editor toolbar option on the Paragraph block. It replaces an `a` tag. This adds an additional icon and the visibly hidden text to the url.
- Use the `ExternalLink` component for the permalink url when editing the post header

Furthermore a change has been made to the preview button in the top toolbar so that is a button element instead of an anchor.

## How has this been tested?
- Manually tested visual appearance of icon
- Manually tested 'Opens in a new window' a11y text using VoiceOver
- Ensured automated tests pass

## Screenshots 
Permalink:
screen shot 2018-07-17 at 11 14 39 am

Preview (visually unchanged, but is now a button element):
screen shot 2018-07-17 at 11 14 56 am

Paragraph Block Link:
screen shot 2018-07-17 at 11 14 50 am

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by talldan** [comment]
@karmatosed Would be great if you could just double-check the screenshots look ok. The only thing I spotted is that the link text color isn't the same for the permalink and the rich text link. I can make a separate issue if you think that's worth making consistent.

File: components/button/style.scss [hunk 1]
```diff
@@ -181,6 +181,15 @@
 		padding: 0 8px 1px;
 		font-size: 11px;
 	}
+
+	.components-external-link__icon {
```
Comment (talldan) [inline components/button/style.scss]: I wasn't entirely sure that these are the right selectors to use, or whether I should somehow specify a class of components-button__icon on the icon.

**Comment by talldan** [comment]
ah - I need to add some tests, particularly on the button element. Will add some asap.

File: components/button/style.scss (hunk 1 above)
Comment (noisysocks) [reply components/button/style.scss]: The class should begin with `components-button__` since the node is in the Button component.

**Comment by noisysocks** [comment]
Very supportive of the changes to the link formatting tool and permalink editor here. It looks great 👍 

I'm not so sure about changing the Preview button, though.

In my mind<sup>1</sup>, the point of the external icon is to shape the user's expectation so that they're not surprised when a new tab opens. This makes total sense since the expectation when clicking a hyperlink is that the browser navigates to a new page in-place. 

But I don't think users have any such expectations when they click on a button. A button doesn't imply that clicking it will navigate to a new page. It's a button—buttons can do anything!

Could an alternative to this be to make the Preview button a `<button>` instead of an `<a>`? Or, could we add we add `aria-role="button"` and have it open when a user selects it and presses Space?

<sup>1</sup> Read: I am not a UX designer! 🙂

**Review by karmatosed — CHANGES_REQUESTED** [review]
@noisysocks is hereby granted temporary 'UX designer' status as he was on point with suggesting this doesn't get done to the preview button :) I would say apart from that so long as things are just links (not buttons hiding in link clothing) this is a great improvement.

**Review by noisysocks — CHANGES_REQUESTED** [review]

**Comment by afercia** [comment]
> and I'm wary of Button becoming a component that has too many responsibilities.

This is the main reason why a previous attempt in #6786 was abandoned. It ended up in adding a lot of stuff to the Button component. Currently, the Button component is named based on its visual appearance. This is far from ideal and introduces some potential confusion. Components should be named based on what they do. In the discussion on https://github.com/WordPress/gutenberg/pull/6786#issuecomment-392903261 at some point it was agreed to split the Button component in two separate components, in short:
- Link: it's an `<a>` element, used to navigate to a web resource (i.e. no `href="#"` links used to "do something") and visually styled as a link or a button
- Button: it's a `<button>` element, used to "do something" and visually styled as a button or link

Based on that #7534 was created to separate the two components _and eliminate ExternalLink_. I know that takes a different path from the work done here, sorry for that, but I'd strongly encourage to focus on the effort proposed on #7534.

**Review by noisysocks — APPROVED** [review]
Aside from the failing test this looks great! 🙂

**Comment by noisysocks** [comment]
It looks like the failing test was just an intermittent E2E test failure. I restarted the job for you.

**Comment by talldan** [comment]
Great - thanks @noisysocks.

@karmatosed - I've updated the screenshots at the top, are you happy to approve this now?

**Comment by karmatosed** [comment]
@talldan do we need to have the link icon in the actual link interface? That feels over the top to me.

**Comment by talldan** [comment]
@karmatosed - do you mean this icon? external link

I thought that was one of the main points we were trying to address from #1105. My opinion is that we should use the icon consistently on anything that looks like and interacts like a link.

**Comment by karmatosed** [comment]
@talldan my understanding was that it was for interface links not links in the drop down. It's this bit I am questioning:

2018-07-18 at 14 12

If everyone wants that's ok it just feels a little odd seeing beside something like this.

**Comment by karmatosed** [comment]
Let's add it in @talldan and iterate.

**Review by karmatosed — APPROVED** [review]

**Pull request merged**